keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,593 @@
+
+##### e2e tests
+
+AppArmor already has [e2e tests][https://github.com/kubernetes/kubernetes/blob/2f6c4f5eab85d3f15cd80d21f4a0c353a8ceb10b/test/e2e_node/apparmor_test.go],
+but the tests are guarded by the `[Feature:AppArmor]` tag and run as part of the
+[containerd E2E features](https://testgrid.k8s.io/sig-node-containerd#node-e2e-features)
+test suite.
+
+Tests will be tagged as `[Feature:AppArmor]` like it is implemented right now,
+but they will be migrated to use the new fields API.
+
+### Failure and Fallback Strategy
+
+There are different scenarios in which applying an AppArmor profile may fail,
+below are the ones we mapped and their outcome once this KEP is implemented:
+
+| Scenario                                                                                           | API Server Result    | Kubelet Result                                                                                                                                        |
+| -------------------------------------------------------------------------------------------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 1) Using localhost or `runtime/default` profile when container runtime does not support AppArmor.  | Pod created          | The outcome is container runtime dependent. In this scenario containers may 1) fail to start or 2) run normally without having its policies enforced. |
+| 2) Using custom or `runtime/default` profile that restricts actions a container is trying to make. | Pod created          | The outcome is workload and AppArmor dependent. In this scenario containers may 1) fail to start, 2) misbehave or 3) log violations.                  |
+| 3) Using a localhost profile that does not exist on the node.                                      | Pod created          | Container runtime dependent: containers fail to start. Retry respecting RestartPolicy and back-off delay.  Error message in event.                |
+| 4) Using an unsupported runtime profile (i.e. `runtime/default-audit`).                            | Fails validation: pod **not** created. | N/A                                                                                                                                                   |
+| 5) Using localhost profile with invalid (empty) name                                               | Fails validation: pod **not** created. | N/A                                                                                                                                                   |
+| 6) AppArmor is disabled by the host or the build                                                   | Pod created. | Kubelet puts Pod in blocked state.                                                                                                                    |
```

Comment:
one thing here is the difference between specifying profile to be `runtime/default` explicitly and having it implicit. When it is explicitly specified to be default, than Pod will be blocked if AppArmor is disabled. If implicit (not specified in pod spec), pod runs as expected.

This difference in behavior is not clearly spelled out in documentation and would be great to highlight.

Reply:
Also I wonder if test for this can be added. I doubt it will be easy as it's hard to disable AppArmor for e2e. So maybe it will only be covered with unit tests